Move from pcomplete to completion-at-point-functions

jabranham opened the pull request:
Explanation in git commit messages (mainly the first one). Please play around and let me know what's broken.

Issue #171

CC'ing people participating there: @DamienCassou @enderw88 @basil-conto @glasser

DamienCassou commented:
I've started using it and things see working fine. Thank you @jabranham.

purcell commented on ledger-complete.el:
```diff
@@ -47,11 +43,19 @@ This file will then be used as a source for account name completions."
   :type 'file
   :group 'ledger)
 
+(defcustom ledger-complete-in-steps t
```
I guess this is here to preserve some of the pcomplete-style behaviour? Because ideally completion in steps would simply be provided (or not) by the active completion system.

jabranham replied:
Yes, though the more I think about it the less it makes sense to preserve. Completion should take care of "completing in steps" if the user wants it that way. So this should probably at least be nil by default, and possibly just removed altogether.

jabranham replied:
I've changed the default to nil.

purcell replied:
Yep, that sounds like a good compromise, thanks!

purcell commented on ledger-complete.el:
```diff
@@ -90,7 +94,7 @@ This file will then be used as a source for account name completions."
                                   payees-list)))))  ;; add the payee
     ;; to the list
-    (pcomplete-uniquify-list (nreverse payees-list))))
+    (delete-dups (sort (nreverse payees-list) #'string-lessp))))
```
If sorting, you don't need to `nreverse it`, right?

purcell commented on test/complete-test.el:
```diff
@@ -35,25 +35,26 @@ http://bugs.ledger-cli.org/show_bug.cgi?id=969
+      (move-end-of-line 1)
+      (newline)
+      (insert "    Dé")
+      (call-interactively #'indent-for-tab-command)
```
I suspect this should call `completion-at-point` instead... `indent-for-tab-command` only completes if `tab-always-indent` is set to `complete`, which it is not by default.

purcell commented on ledger-mode.el:
```diff
@@ -330,17 +317,15 @@ With a prefix argument, remove the effective date."
   (ledger-init-load-init-file)
   (setq-local comment-start ";")
-
+  (setq-local tab-always-indent 'complete)
```
This is a user-level customization option which you shouldn't override in a major mode. If you're worried about users finding their tab completion suddenly broken, we could (for a while) print a message if `tab-always-indent` is not set to this value.

jabranham replied:
I think it makes sense to set tab-always-indent to 'complete here, since that preserves ledger-mode's previous behavior. 

That said, it's not at all a hill I'm willing to die on, so I've gone ahead and removed this bit.

purcell commented on ledger-mode.el:
```diff
@@ -330,17 +317,15 @@ With a prefix argument, remove the effective date."
   (setq-local comment-start ";")
-
+  (setq-local tab-always-indent 'complete)
+  (setq-local indent-line-function #'ledger-indent-line)
```
Nice.

purcell commented:
Also probably need a couple of little doc changes here.

jabranham commented:
@purcell Thanks for the review! I think I addressed all your comments, let me know if I missed anything. I looked through the texi manual and added a bit about tab completion. Any other doc changes you think are needed?

purcell commented:
> Any other doc changes you think are needed?

No, that looks great, thanks. While I'm very happy with this PR, I'll hold off on merging this myself until @enderw88 has taken a look too.

enderw88 commented:
I am camping for the week and won’t be able to  look until next Sunday. If
you are happy with it merge it.

On Mon, May 27, 2019 at 13:48 Steve Purcell <notifications@github.com>
wrote:

> Any other doc changes you think are needed?
>
> No, that looks great, thanks. While I'm very happy with this PR, I'll hold
> off on merging this myself until @enderw88 <https://github.com/enderw88>
> has taken a look too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ledger/ledger-mode/pull/174?email_source=notifications&email_token=AAGWWLDLECPB2NU764D6TPTPXRCH7A5CNFSM4HPWWS72YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWKPXOY#issuecomment-496303035>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGWWLD66R7T65CBGUWQIK3PXRCH7ANCNFSM4HPWWS7Q>
> .
>
-- 
Craig, Corona De Tucson, AZ
[image: missile_flyout]
enderw88.wordpress.com

jwiegley commented:
Does this drop the ability to cycle through completions by hitting TAB?

jabranham commented:
@jwiegley Yes, it does. Instead it respects `completion-at-point-functions`. Perhaps there is a pcomplete style that users can add to `completion-styles` to re-enable the old behavior?

I tried at first to keep the pcompletion stuff in there but it became too complicated.

jwiegley commented:
The cycling is the #1 reason I wrote pcomplete in the first place. Excising it from ledger-mode would be most unfortunate. Most completions I want to do are within 1 or 2 tabs, rather than having to always deal with popups or selecting from a list.

jabranham commented:
OK, after a bit of experimenting/learning about Emacs's other completion options, it appears you can get the cyclical completion back with:

```elisp
(add-hook 'ledger-mode-hook
          (lambda ()
            (setq-local tab-always-indent t)
            (setq-local completion-cycle-threshold t)
            (setq-local ledger-complete-in-steps t)))
```

purcell commented:
Great, I've added a commit which notes this in the docs, and will go ahead and merge shortly.

basil-conto commented on ledger-complete.el:
```diff
@@ -223,37 +227,32 @@ Looks in `ledger-accounts-file' if set, otherwise the current buffer."
+           (progn (setq start (save-excursion (back-to-indentation) (point)))
+                  (if ledger-complete-in-steps
+                      (setq collection #'ledger-accounts-tree)
+                    (setq collection #'ledger-accounts-list-in-buffer)))))
```
Nitpick:

```el
(setq collection (if ledger-complete-in-steps
                     #'ledger-accounts-tree
                   #'ledger-accounts-list-in-buffer))

bcc32 replied:
I think this also breaks the `ledger-accounts-file` option.  You probably want `ledger-accounts-list` instead.

jabranham replied:
@bcc32 thanks. I noticed one other small thing and opened a PR for it and this.

basil-conto commented on ledger-post.el:
```diff
@@ -120,6 +120,18 @@ Looks only as far as END, if supplied, otherwise `point-max'."
+  "Indent the current line."
+  (if (save-excursion
+        (forward-line -1)
+        (beginning-of-line)
```
No need for `beginning-of-line`, `forward-line` already leaves point at BOL (unless you're doing something complicated with field restrictions).

basil-conto commented on ledger-post.el:
```diff
@@ -120,6 +120,18 @@ Looks only as far as END, if supplied, otherwise `point-max'."
+  (if (save-excursion
+        (forward-line -1)
+        (beginning-of-line)
+        (not (looking-at-p (rx bol (0+ (or "\n" whitespace)) eol))))
```
Tip: If you remove the `not`, then you can swap the next two arguments to `if` around, and spare yourself the `progn`.

basil-conto commented on ledger-post.el:
```diff
@@ -120,6 +120,18 @@ Looks only as far as END, if supplied, otherwise `point-max'."
+        (forward-line -1)
+        (beginning-of-line)
+        (not (looking-at-p (rx bol (0+ (or "\n" whitespace)) eol))))
+      (progn (when (not (= ledger-post-account-alignment-column
```
Tip: `when`+`not` = `unless`.

basil-conto commented on test/complete-test.el:
```diff
@@ -35,25 +35,26 @@ http://bugs.ledger-cli.org/show_bug.cgi?id=969
-    (equal (buffer-string)
-           "2013/05/19 Retrait
+      (forward-line 1)
+      (move-end-of-line 1)
```
Off-topic: Does this really have to be the interactive command `move-end-of-line`, or would `end-of-line` do? Either way, why call `forward-line` when both `move-end-of-line` and `end-of-line` accept as argument a number of lines to move forward?

basil-conto commented on test/mode-test.el:
```diff
@@ -61,27 +61,19 @@
+    (comment-dwim nil)
+    (should (string-match (rx buffer-start ";" (0+ whitespace))
+                          ;; Expected: no space before ';'
+                          (buffer-string)))))
```
Nitpick: `string-match-p`? Alternatively, you could forgo allocating `buffer-string` by going to BOB and call `looking-at-p`.

basil-conto commented on test/mode-test.el:
```diff
@@ -61,27 +61,19 @@
+  (ledger-tests-with-temp-file ""
+    (setq comment-start "#")
+    (comment-dwim nil)
+    (should (string-match (rx buffer-start "#" (0+ whitespace)) (buffer-string)))))
```
Ditto re: `string-match-p`/`looking-at-p`.

purcell commented:
Thanks @basil-conto -- I just pushed fixes for most of these nits, at least in the `ledger-*.el` files.

jabranham commented:
@basil-conto thanks for the review and thanks @purcell for taking care of it!

jaccarmac commented:
Might be missing something, but after this change TAB no longer indents amounts for me. The section in the manual just below the changed part suggests that I should set `ledger-post-auto-adjust-amounts`, but that doesn't seem to be used anywhere in the code base. How can I get amount justification back with the new completion system?

jaccarmac commented:
In addition, `complete-in-steps` seems not to be respected for me.

jabranham commented:
Have you set tab-always-indent to 'complete?

On May 31, 2019 2:49:48 AM CDT, Jacob MacDonald <notifications@github.com> wrote:
>Might be missing something, but after this change TAB no longer indents
>amounts for me. The section in the manual just below the changed part
>suggests that I should set `ledger-post-auto-adjust-amounts`, but that
>doesn't seem to be used anywhere in the code base. How can I get amount
>justification back with the new completion system?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/ledger/ledger-mode/pull/174#issuecomment-497612633

jaccarmac commented:
Yep, using the snippet from the manual as you can verify [in my config](http://chiselapp.com/user/jaccarmac/repository/dot-emacs-dot-d/info/8528e16ddf728f86).

brabalan commented:
I'm also having issues since this was merged:
- I cannot complete an account that is already reconciled (i.e., if I type `* Exp` there is no completion, although it should complete to `* Expenses:`), removing the star works, but I need to `C-c C-c` the line after, which is a burden
- initial completion does not care about case (`exp` completes to `Expenses:`), but further completion cares about case (I suspect this is because the completed part has an upper case letter, so case matters)

bcc32 commented:
One other random thing I noticed: for payee completion, the incomplete input is added to the list of suggested completions, which seems like the wrong behavior.  For example, if I type `Wal<tab>` (expecting completion to give me just `Walmart`), I am instead asked to choose between `Wal` and `Walmart`.

jabranham commented:
Thanks. I'm busy until Friday but should have time to take a look at these issues then

On June 2, 2019 12:51:28 PM EST, Aaron Zeng <notifications@github.com> wrote:
>One other random thing I noticed: for payee completion, the incomplete
>input is added to the list of suggested completions, which seems like
>the wrong behavior.  For example, if I type `Wal<tab>` (expecting
>completion to give me just `Walmart`), I am instead asked to choose
>between `Wal` and `Walmart`.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/ledger/ledger-mode/pull/174#issuecomment-498051543

cpitclaudel commented:
I think this broke company integration, too.  Before, if I had two accounts `AA:BB:CC` and `AA:BB:DD`, typing `AA:B` then `RET` would insert `AA:BB:`, and then pressing `company-manual-begin` would resume completion and offer `CC` and `DD`.  But now a space is inserted after `AA:BB: `.

enderw88 commented:
I am feeling like a really crochety old man. Over the last few years there
have been many very well meaning people trying to improve ledger-mode and
mostly they have fixed a nit then  caused a crater.

Every “improvement” in the last few years seems to have seriously disrupted
existing workflows.   Do we need an “experimental” branch for people who
want the bleeding edge?

I am tempted to go private with my fork and keep it working for me.

On Sun, Jun 2, 2019 at 17:24 Clément Pit-Claudel <notifications@github.com>
wrote:

> I think this broke company integration, too. Before, if I had two accounts
> AA:BB:CC and AA:BB:DD, typing AA:B then RET would insert AA:BB:, and then
> pressing company-manual-begin would resume completion and offer CC and DD.
> But now a space is inserted after AA:BB: .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ledger/ledger-mode/pull/174?email_source=notifications&email_token=AAGWWLE2IEV3J4DXNYR2JH3PYRQEDA5CNFSM4HPWWS72YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWYBFUY#issuecomment-498078419>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGWWLEFTXZKUH6WAXE2ELTPYRQEDANCNFSM4HPWWS7Q>
> .
>
-- 
Craig, Corona De Tucson, AZ
[image: missile_flyout]
enderw88.wordpress.com

cpitclaudel commented:
>  Every “improvement” in the last few years seems to have seriously disrupted existing workflows.

I think that's an overstatement :) There's been plenty of bugfixes, for example, and many actual improvements, too (of course, ledger-mode was already pretty good to begin with; thanks @enderw88!).

> Do we need an “experimental” branch for people who want the bleeding edge?

That's what `master` is; people who do not want the bleeding edge are supposed to use MELPA stable, instead of MELPA.  Was this change already released MELPA stable?  If not, things are mostly working as they should: bugs are being ironed out in preparation for the next release.

> I am tempted to go private with my fork and keep it working for me.

It's easy to get frustrated, but in my experience many projects are already stretched thin, and it can be painful to read this type of comments.  In many (most?) cases, getting an imperfect contribution is well worth it if it means also onboarding a new contributor, or simplifying a hard-to-maintain part of the code, or setting the stage for further improvements, etc.

That being said, if that frustration leads you to resume active work on ledger-mode, I'm sure it'll be a win for everyone :)

enderw88 commented:
You are probably right. Like I said, I am being a crochety old man. I
appreciate the kind words.

I use Master as my day-to-day so I can see if there are major breakages
before they flow out.  But I don't deal with the EMACS distribution stream.

On Sun, Jun 2, 2019 at 19:58 Clément Pit-Claudel <notifications@github.com>
wrote:

> Every “improvement” in the last few years seems to have seriously
> disrupted existing workflows.
>
> I think that's an overstatement :) There's been plenty of bugfixes, for
> example, and many actual improvements, too (of course, ledger-mode was
> already pretty good to begin with; thanks @enderw88
> <https://github.com/enderw88>!).
>
> Do we need an “experimental” branch for people who want the bleeding edge?
>
> That's what master is; people who do not want the bleeding edge are
> supposed to use MELPA stable, instead of MELPA. Was this change already
> released MELPA stable? If not, things are mostly working as they should:
> bugs are being ironed out in preparation for the next release.
>
> I am tempted to go private with my fork and keep it working for me.
>
> It's easy to get frustrated, but in my experience many projects are
> already stretched thin, and it can be painful to read this type of
> comments. In many (most?) cases, getting an imperfect contribution is well
> worth it if it means also onboarding a new contributor, or simplifying a
> hard-to-maintain part of the code, or setting the stage for further
> improvements, etc.
>
> That being said, if that frustration leads you to resume active work on
> ledger-mode, I'm sure it'll be a win for everyone :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ledger/ledger-mode/pull/174?email_source=notifications&email_token=AAGWWLGU74FNJWECYMFQSZTPYSCGLA5CNFSM4HPWWS72YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWYFV6Y#issuecomment-498096891>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGWWLBAISVWHBJ7JW4B4Q3PYSCGLANCNFSM4HPWWS7Q>
> .
>

jwiegley commented:
I'm OK with reverting this PR until we find a satisfying answer for everyone, btw.

purcell commented:
> That's what `master` is; people who do not want the bleeding edge are supposed to use MELPA stable, instead of MELPA. Was this change already released MELPA stable? If not, things are mostly working as they should: bugs are being ironed out in preparation for the next release.

This is my position too: let's be conservative in tagging stable releases, while tolerating a little short-term change and breakage on master in the name of modernising the code and making it more maintainable.

There are so many different usage styles for this code and they're not captured in tests, so these recent instances of breakage couldn't reasonably have been avoided by well-meaning contributors: if anything, this is an opportunity to add more tests.

enderw88 commented:
I need to learn how to add behavioral tests. I don’t have any idea how to
implement a test that would show tab completion cycles appropriately.

On Mon, Jun 3, 2019 at 15:57 Steve Purcell <notifications@github.com> wrote:

> That's what master is; people who do not want the bleeding edge are
> supposed to use MELPA stable, instead of MELPA. Was this change already
> released MELPA stable? If not, things are mostly working as they should:
> bugs are being ironed out in preparation for the next release.
>
> This is my position too: let's be conservative in tagging stable releases,
> while tolerating a little short-term change and breakage on master in the
> name of modernising the code and making it more maintainable.
>
> There are so many different usage styles for this code and they're not
> captured in tests, so these recent instances of breakage couldn't
> reasonably have been avoided by well-meaning contributors: if anything,
> this is an opportunity to add more tests.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ledger/ledger-mode/pull/174?email_source=notifications&email_token=AAGWWLBXY6W63HLEYXUQ45TPYWOUBA5CNFSM4HPWWS72YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODW25M3A#issuecomment-498456172>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGWWLAD4JHHGBQHAY6NRDDPYWOUBANCNFSM4HPWWS7Q>
> .
>
-- 
Craig, Corona De Tucson, AZ
[image: missile_flyout]
enderw88.wordpress.com

jaccarmac commented:
I'm with @cpitclaudel and @purcell: Breakage is expected.

In this case it just caught my attention because there were two PRs about the same topic. I regret not being familiar enough with completion-at-point to offer a patch instead of a complaint off the bat.

purcell commented:
> (@jaccarmac) Might be missing something, but after this change TAB no longer indents amounts for me. The section in the manual just below the changed part suggests that I should set `ledger-post-auto-adjust-amounts`, but that doesn't seem to be used anywhere in the code base. How can I get amount justification back with the new completion system?

I made some further tweaks in 2c0c68af, including that amounts are (unconditionally) re-aligned post-completion. I think that `ledger-indent-line` still needs a bit of work, and I'll look into making the code respect `ledger-post-auto-adjust-amounts`, which option seems to have been lost in the sands of time.

> (@jaccarmac) In addition, `complete-in-steps` seems not to be respected for me.

Oh. It seemed to be working for me with the above changes. Perhaps you could share a quick test case in a comment here?

purcell commented:
> I'll look into making the code respect `ledger-post-auto-adjust-amounts`, which option seems to have been lost in the sands of time.

In fact, that var isn't even defined any more, so I'd suggest the best resolution is to remove the docs for it.

jaccarmac commented:
The minimal case is simpler than I thought.

```
2019/07/05 One.
    Assets:Foo   $1
    Assets:Bar  -$1
```

Start a new transaction as so.

```
2019/07/05 Two.
    A
```

Then press `TAB`. I would expect "Assets" to complete but instead "Assets:Bar" appears and I can cycle to "Assets:Foo". This is more eager behavior than with the old tab-completion. Interestingly, if you enter a transaction without the "A" and press `TAB` only the first segment of the account will be completed.

This may be related to the issue @cpitclaudel reported, though I don't have any three-segment account names in my file to have noticed.

jaccarmac commented:
And thanks for the doc-update. It still doesn't align to the column in question for me but I'll chock that up to new bugs and not a variable that doesn't exist any more :+1:

purcell commented:
Hmm, thanks. That case is actually working okay for me now with the code now in `master`. (I just pushed some additional fixes.)

purcell commented:
(The fixes include an overhaul of `ledger-indent-line`, which should now correctly trigger automatic realignment.)

jaccarmac commented:
Sweet, indentation does seem to work now, even better than before (only one `TAB` press needed). However, neither indentation nor completion works unless the line is manually indented 4 spaces.

purcell commented:
Sorry, I'm a bit confused: you say indentation works, but _indentation_ doesn't work unless the line is indented? If you could describe brief steps to reproduce, that would be helpful, because I'm finding both indentation and completion work okay now. (Note that account name completion only kicks in now on indented lines, whereas before it was a bit of a "catch-all" completion when no other contextual completion applied: I think this makes sense, but please correct me if I'm wrong, anyone.)

purcell commented:
Perhaps you were using TAB characters for indentation, in which case I've made a small fix, sorry.

jaccarmac commented:
Nah, I was using four spaces. Basically, if I pressed `RET` or `C-j` from the end of a new transaction line (`C-c C-a` then typing some text) the cursor would be in column 0 and pressing `TAB` would do nothing. If I then indented by 4 spaces completion and indentation of the second column would work, and `RET` would take me to column 4 in the next line with 4 spaces inserted. Haven't tried the latest patch yet, will do so.

jaccarmac commented:
Still seeing the same behavior on master. Minimal case is again minimal.

```
2019/06/06 Test
```

Put your cursor at the end of the line and press `RET`

purcell commented:
Thanks, fixed in dfd6d0ef.

jabranham commented:
Thanks @purcell for cleaning up after this! Like I said above, I should have time now to make fixes, but from my reading you seem to have beaten me to it. Let me know if there are any other outstanding issues and I can take a look.

purcell commented:
@jabranham No worries. There are indeed a couple more outstanding issues I noticed:

- Date completion appears broken, in that if I type "20" at the beginning of the line and hit TAB, I get a backtrace: `Lisp error: (wrong-number-of-arguments (3 . 3) 0)
  #f(compiled-function (string predicate all) #<bytecode 0x4049d0a1>)()`. Even when I hacked it a bit locally to get it roughly working, it doesn't feel great to use, so could use a little love.
- Metadata completion is broken (though this isn't a regression due to this PR, it seems): see  #166.

jabranham commented:
@purcell 

I'm happy to work on this, particularly date completion, but I'm not sure what date completion should look like. Do we just offer all previously-typed dates? Or all dates within +/- a month?

purcell commented:
> I'm happy to work on this, particularly date completion, but I'm not sure what date completion should look like. Do we just offer all previously-typed dates? Or all dates within +/- a month?

There's a regexp for "partial" dates, which I think is a bit bogus anyway, but I think the ideal is that the possible month and day portions of a date should get completed within the current year, or year+month if specified. (Generating 365 or 366 completions would be the worst case.) Handling the variable ledger date formats is the tricky bit here, and I'm not confident the existing code worked well for this.

jabranham commented:
I don't think date completion ever worked for me (though I usually add a new transaction with C-c C-a anyway).

Maybe we can take this opportunity to include the other things that can start a line too (e.g. account).

On Fri 14 Jun 2019 at 20:54, Steve Purcell <notifications@github.com> wrote:

>> I'm happy to work on this, particularly date completion, but I'm not sure what date completion should look like. Do we just offer all previously-typed dates? Or all dates within +/- a month?
>
> There's a regexp for "partial" dates, which I think is a bit bogus anyway, but I think the ideal is that the possible month and day portions of a date should get completed within the current year, or year+month if specified. (Generating 365 or 366 completions would be the worst case.) Handling the variable ledger date formats is the tricky bit here, and I'm not confident the existing code worked well for this.
